opacus/tests/privacy_engine_validation_test.py

```diff
@@ -0,0 +1,216 @@
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from opacus import PrivacyEngine
+from torch.utils.data import DataLoader
+
+
+class BasicSupportedModule(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.conv = nn.Conv1d(in_channels=16, out_channels=8, kernel_size=2)
+        self.gn = nn.GroupNorm(num_groups=2, num_channels=8)
+        self.fc = nn.Linear(in_features=4, out_features=8)
+        self.ln = nn.LayerNorm([8, 8])
+
+    def forward(self, x):
+        x = self.conv(x)
+        x = self.gn(x)
+        x = self.fc(x)
+        x = self.ln(x)
+        return x
+
+
+class CustomLinearModule(nn.Module):
+    def __init__(self, in_features, out_features):
+        super().__init__()
+        self._weight = nn.Parameter(torch.empty(out_features, in_features))
+        self._bias = nn.Parameter(torch.empty(out_features))
+
+    def forward(self, x):
+        return F.linear(x, self._weight, self._bias)
+
+
+class MatmulModule(nn.Module):
+    def __init__(self, input_features, output_features):
+        super().__init__()
+        self.weight = nn.Parameter(torch.empty(input_features, output_features))
+
+    def forward(self, x):
+        return torch.matmul(x, self.weight)
+
+
+class LinearWithExtraParam(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.fc = nn.Linear(5, 8)
+        self.extra_param = nn.Parameter(torch.empty(8, 2))
+
+    def forward(self, x):
+        x = self.fc(x)
+        x = x.matmul(self.extra_param)
+        return x
+
+
+class PrivacyEngineValidationTest(unittest.TestCase):
+    """
+    This test case checks end-to-end model validation performed in `.make_private`
+    method. It covers performed in `ModuleValidator`, `GradSampleModule`, as well as
+    their interplay
+    """
+
+    def setUp(self) -> None:
+        self.privacy_engine = PrivacyEngine()
+
+    def _init(self, module):
+        optim = torch.optim.SGD(module.parameters(), lr=0.1)
+        dl = DataLoader(
+            dataset=[torch.randn(16, 5) for _ in range(100)],
+            batch_size=10,
+        )
+
+        return module, optim, dl
+
+    def test_supported_hooks(self):
+        module, optim, dl = self._init(BasicSupportedModule())
+
+        module, optim, dl = self.privacy_engine.make_private(
+            module=module,
+            optimizer=optim,
+            data_loader=dl,
+            noise_multiplier=1.0,
+            max_grad_norm=1.0,
+            grad_sample_mode="hooks",
+        )
+
+        for x in dl:
+            module(x)
+
+    def test_supported_ew(self):
+        module, optim, dl = self._init(BasicSupportedModule())
+
+        module, optim, dl = self.privacy_engine.make_private(
+            module=module,
+            optimizer=optim,
+            data_loader=dl,
+            noise_multiplier=1.0,
+            max_grad_norm=1.0,
+            grad_sample_mode="ew",
+        )
+
+        for x in dl:
+            module(x)
```

Comment:
nit: how about just one test with 
```
@given(
  grad_sample_mode=st.sampled_from(["ew","hooks"]),
)
```
?
Not a blocker for this PR at all though.

Reply:
Oh, yes, sorry forgot to respond here.

To be honest, I'm not a big fan of using `hypothesis` for tests where we need to test two branches. Hypothesis samples the parameters randomly. In combination with a limited number of runs, it doesn't actually guarantee that any particular set of parameters will be executed every time you run tests. 
In reality this problem would only surface if the function has a lot of independent parameters, but nevertheless it doesn't look like a good practice to me.

I'd say it'l ok to use to to sample things like dimensions, but if we want to test different flags or implementations, I'd rather use 'unittest.subTest'